File: layoutmorphism.go
package tensorfs

// LAYOUT MORPHISMS: where a tensor's elements SIT IN STORAGE, as data.
//
// A topology says WHICH tensors exist and what shape they are; a quant rule
// says what element type they are stored in. Neither says anything about the
// ORDER the elements are written in — and that order is where quantized fast
// paths and custom kernels live. Today every layer repacks by hand: inductor
// re-derives NHWC conv weights every forward call, nunchaku's loader unswizzles
// seven sub-axes of a scale tensor in Python, and two byte-different packagings
// of one checkpoint are stored twice because nothing can say they are the same
// tensors in a different arrangement.
//
// A layout morphism says it, in one closed form:
//
//	FACTOR the logical shape into sub-axes, PERMUTE the sub-axes, write
//	row-major. A sub-axis product that EXCEEDS its logical dim is padding.
//
// That is the whole language, and every entry in the v1 vocabulary is one
// instance of it:
//
//	contiguous            the identity — no factorization, no permutation
//	channels_last (2d/3d) NCHW -> NHWC, NCDHW -> NDHWC
//	transposed            a rank-2 swap
//	stride-padding        the innermost dim rounded up to an alignment
//	cublas.blockscale     128x4-tiled nvfp4 block scales (padded AND blocked)
//	nunchaku.micro-scale  a seven-sub-axis warp-lane interleave
//
// WHY THIS FORM AND NOT CODE. A morphism written as code is ratified by
// review, and review is exactly what missed nunchaku's `(4, 4, 8)` row split
// reshaping without error into `(4, 8, 4)` — every name, dtype and shape
// correct and every channel silently permuted. Written as data, the map is
// MACHINE-VERIFIABLE: apply it and apply its inverse and compare bytes. A
// record that survives that is AUTO-RATIFIED; no human signs a permutation.
//
// THE TRANSFORM ITSELF IS NOT HERE. This file decides — parses, validates,
// prices and plans. The bytes are moved in exactly one place, the Rust crate's
// `layout_morphism` module, which has two backends (CPU materialization and
// the GPU fill) over ONE implementation. A Go copy of the applier would be a
// second answer to "which byte goes where", which is the disease v1's three
// matchers already proved is undetectable until a pod serves noise.

import (
	"bytes"
	"encoding/json"
	"fmt"
	"strings"
)

// LayoutMorphismFormat tags the record class.
const LayoutMorphismFormat = "tensorfs-layout-morphism-v2"

// DefaultLayout is the byte arrangement every tree carries unless it says
// otherwise: plain row-major. It is the identity morphism, and it is the
// coordinate every stamp written before this file existed means.
var DefaultLayout = Handle{Name: "torch.contiguous", Version: 1}

// LayoutClass sizes a morphism, and the two classes are worth different money.
type LayoutClass string

const (
	// ClassInductor is the compiler's own closed vocabulary. Worthwhile but
	// small: inductor races kernels, not layouts, and its layout pass is
	// rule-based.
	ClassInductor LayoutClass = "inductor"
	// ClassEndpointDeclared is the kernel-library class — the arrangement a
	// quantized fast path or a custom kernel demands. These are the real wins.
	ClassEndpointDeclared LayoutClass = "endpoint-declared"
)

// SubAxis is one factor of one logical axis.
//
// Extent is a formula in the ONE shape language v2 already has (`d0`, `d1/16`,
// literals, `ceil(...)`), so a padded factorization spells its rounding the
// same way a quant rule's emission does. That is not a convenience: the cuBLAS
// block-scale layout's element count and `bfl.nvfp4-preswizzled@1`'s emission
// shape are two documents that must agree, and they agree in one language.
type SubAxis struct {
	Axis   int    `json:"axis"`
	Extent string `json:"extent"`
}

// LayoutMorphism is one validated storage arrangement.
type LayoutMorphism struct {
	handle      Handle
	class       LayoutClass
	description string
	// rank pins the logical rank this arrangement is defined for. 0 means any
	// rank, which only the identity can honestly claim.
	rank    int
	subAxes []SubAxis
	extents []*shapeExpr
	// perm is the STORAGE order of the sub-axis list: perm[j] is the index of
	// the sub-axis that varies at storage position j.
	perm []int
	// candidate marks a wish emitted by a compiler that has not been ratified.
	// A candidate is parsed, validated and verified like any other record and
	// still NEVER derives: machines derive along ratified morphisms, they do
	// not invent them.
	candidate  bool
	provenance []string
}

func (m *LayoutMorphism) Handle() Handle       { return m.handle }
func (m *LayoutMorphism) Class() LayoutClass   { return m.class }
func (m *LayoutMorphism) Description() string  { return m.description }
func (m *LayoutMorphism) Rank() int            { return m.rank }
func (m *LayoutMorphism) SubAxes() []SubAxis   { return m.subAxes }
func (m *LayoutMorphism) Permutation() []int   { return append([]int(nil), m.perm...) }
func (m *LayoutMorphism) Candidate() bool      { return m.candidate }
func (m *LayoutMorphism) Ratified() bool       { return !m.candidate }
func (m *LayoutMorphism) Provenance() []string { return m.provenance }
func (m *LayoutMorphism) Digest() string       { return digestOf(m.canonical()) }
func (m *LayoutMorphism) Identity() bool       { return len(m.subAxes) == 0 }
func (m *LayoutMorphism) String() string       { return m.handle.String() }

func (m *LayoutMorphism) canonical() string {
	var out strings.Builder
	out.WriteString(LayoutMorphismFormat)
	fmt.Fprintf(&out, "\nname=%s\nversion=%d\nclass=%s\nrank=%d\ncandidate=%t\n",
		m.handle.Name, m.handle.Version, m.class, m.rank, m.candidate)
	for _, sub := range m.subAxes {
		fmt.Fprintf(&out, "sub axis=%d extent=%s\n", sub.Axis, sub.Extent)
	}
	rendered := make([]string, 0, len(m.perm))
	for _, at := range m.perm {
		rendered = append(rendered, fmt.Sprint(at))
	}
	fmt.Fprintf(&out, "perm=%s\n", strings.Join(rendered, ","))
	return out.String()
}

type rawLayoutMorphism struct {
	Format      *string   `json:"format"`
	Name        *string   `json:"name"`
	Version     *uint32   `json:"version"`
	Class       string    `json:"class"`
	Description string    `json:"description"`
	Rank        int       `json:"rank,omitempty"`
	SubAxes     []SubAxis `json:"sub_axes,omitempty"`
	Permutation []int     `json:"permutation,omitempty"`
	Candidate   bool      `json:"candidate,omitempty"`
	Provenance  []string  `json:"provenance"`
	Digest      string    `json:"digest,omitempty"`
}

// ParseLayoutMorphism reads and validates one layout record.
//
// Everything checkable WITHOUT a shape is checked here; everything that needs
// one is checked by Plan. The split matters: a record is loaded once at start
// and planned per tensor, so a malformed permutation must not wait for a
// checkpoint to arrive before it refuses.
func ParseLayoutMorphism(document []byte) (*LayoutMorphism, error) {
	decoder := json.NewDecoder(bytes.NewReader(document))
	decoder.DisallowUnknownFields()
	var raw rawLayoutMorphism
	if err := decoder.Decode(&raw); err != nil {
		return nil, refuse("json", "%v", err)
	}
	if raw.Format == nil || *raw.Format != LayoutMorphismFormat {
		return nil, refuse("format", "not a %s document", LayoutMorphismFormat)
	}
	if raw.Name == nil || raw.Version == nil {
		return nil, refuse("identity", "a layout morphism is named and versioned")
	}
	handle, err := ParseHandle(fmt.Sprintf("%s@%d", *raw.Name, *raw.Version))
	if err != nil {
		return nil, err
	}
	class := LayoutClass(raw.Class)
	if class != ClassInductor && class != ClassEndpointDeclared {
		return nil, refuse("layout-class",
			"%s declares class %q; a layout is either the compiler's (%q) or a "+
				"kernel library's (%q)", handle, raw.Class, ClassInductor, ClassEndpointDeclared)
	}
	if len(raw.Provenance) == 0 {
		return nil, refuse("provenance",
			"%s carries no evidence. A layout is transcribed from the code that "+
				"reads these bytes; the citation is the only reason to trust it", handle)
	}
	morphism := &LayoutMorphism{
		handle: handle, class: class, description: raw.Description, rank: raw.Rank,
		subAxes: raw.SubAxes, perm: raw.Permutation, candidate: raw.Candidate,
		provenance: raw.Provenance,
	}
	if err := morphism.validate(); err != nil {
		return nil, err
	}
	if raw.Digest != "" && raw.Digest != morphism.Digest() {
		return nil, refuse("digest", "%s carries digest %s but hashes to %s",
			handle, raw.Digest, morphism.Digest())
	}
	return morphism, nil
}

func (m *LayoutMorphism) validate() error {
	if len(m.subAxes) == 0 {
		// The identity. It is a real record — it is the coordinate every
		// existing stamp carries — and it is the ONLY one that may be rankless.
		if len(m.perm) != 0 {
			return refuse("layout", "%s permutes a factorization it does not declare", m.handle)
		}
		if m.rank != 0 {
			return refuse("layout",
				"%s factors nothing, which is the identity for EVERY rank; pinning "+
					"it to rank %d would refuse tensors it arranges correctly",
				m.handle, m.rank)
		}
		return nil
	}
	if m.rank < 1 {
		return refuse("layout",
			"%s factors %d sub-axes without declaring the rank they factor",
			m.handle, len(m.subAxes))
	}
	if len(m.perm) != len(m.subAxes) {
		return refuse("layout", "%s permutes %d positions over %d sub-axes",
			m.handle, len(m.perm), len(m.subAxes))
	}
	// Sub-axes are grouped by axis, axes ascending, every axis covered. The
	// order WITHIN a group is the row-major order of the logical axis, so it is
	// load-bearing and cannot be sorted here.
	axis, covered := -1, 0
	for _, sub := range m.subAxes {
		switch sub.Axis {
		case axis: // another factor of the axis already open
		case axis + 1:
			axis, covered = sub.Axis, covered+1
		default:
			return refuse("layout",
				"%s lists axis %d after axis %d; sub-axes are grouped by axis, "+
					"outermost first, and no axis is skipped", m.handle, sub.Axis, axis)
		}
		expression, err := parseShapeExpr(sub.Extent)
		if err != nil {
			return refuse("layout", "%s: axis %d: %v", m.handle, sub.Axis, err)
		}
		m.extents = append(m.extents, expression)
	}
	if covered != m.rank {
		return refuse("layout", "%s declares rank %d and factors %d axes",
			m.handle, m.rank, covered)
	}
	seen := make([]bool, len(m.perm))
	for _, at := range m.perm {
		if at < 0 || at >= len(m.perm) {
			return refuse("layout", "%s permutes to position %d, which does not exist",
				m.handle, at)
		}
		if seen[at] {
			return refuse("layout",
				"%s sends two sub-axes to storage position %d. A layout that is not "+
					"a bijection loses elements, and it loses them silently", m.handle, at)
		}
		seen[at] = true
	}
	if m.changesNothing() {
		return refuse("layout",
			"%s is the identity for every shape — it factors the logical axes, "+
				"permutes nothing and pads nothing. Storing it as a second name for "+
				"%s would make one arrangement two identities", m.handle, DefaultLayout)
	}
	return nil
}

// changesNothing reports a record that is the identity map at EVERY shape:
// storage order equals declaration order and no extent can round up.
func (m *LayoutMorphism) changesNothing() bool {
	for at, to := range m.perm {
		if at != to {
			return false
		}
	}
	for _, sub := range m.subAxes {
		if strings.Contains(sub.Extent, "ceil(") {
			return false
		}
	}
	return true
}

// --- planning one arrangement over one concrete shape -----------------------

// LayoutPlan is what a layout morphism MEANS for one tensor: the storage
// extents, in storage order, and how many elements the destination holds.
//
// It is a decision, not a transform. The Rust applier derives the same extents
// from the same record and moves the bytes; nothing here writes one.
type LayoutPlan struct {
	// Source is the logical shape this plan was computed for.
	Source Shape
	// Extents are the sub-axis extents in DECLARATION order.
	Extents []uint64
	// Storage are the same extents in STORAGE order (extents[perm[j]]).
	Storage []uint64
	// Elements is the destination element count: the product of Storage.
	Elements uint64
	// Padded reports that the destination is LARGER than the source, because
	// some axis rounded up. A padded layout is injective, not bijective: it
	// round-trips on its image and the fill bytes are zero and meaningless.
	Padded bool
}

// Plan computes this arrangement for a concrete logical shape, or refuses.
func (m *LayoutMorphism) Plan(shape Shape) (*LayoutPlan, error) {
	if m.rank != 0 && len(shape) != m.rank {
		return nil, refuse("layout-rank",
			"%s arranges rank-%d tensors; this one is rank %d %s",
			m.handle, m.rank, len(shape), shape)
	}
	plan := &LayoutPlan{Source: shape.clone(), Elements: 1}
	if m.Identity() {
		plan.Extents = shape.clone()
		plan.Storage = shape.clone()
		for _, dimension := range shape {
			plan.Elements *= dimension
		}
		return plan, nil
	}
	product := uint64(1)
	axis := m.subAxes[0].Axis
	for at, sub := range m.subAxes {
		if sub.Axis != axis {
			if err := m.closeAxis(axis, product, shape); err != nil {
				return nil, err
			}
			if product > shape[axis] {
				plan.Padded = true
			}
			axis, product = sub.Axis, 1
		}
		extent, err := m.extents[at].eval(shape)
		if err != nil {
			return nil, refuse("layout", "%s: axis %d: %v", m.handle, sub.Axis, err)
		}
		if extent == 0 {
			return nil, refuse("layout", "%s: axis %d has a zero-sized factor %q",
				m.handle, sub.Axis, sub.Extent)
		}
		product *= extent
		plan.Extents = append(plan.Extents, extent)
		plan.Elements *= extent
	}
	if err := m.closeAxis(axis, product, shape); err != nil {
		return nil, err
	}
	if product > shape[axis] {
		plan.Padded = true
	}
	for _, at := range m.perm {
		plan.Storage = append(plan.Storage, plan.Extents[at])
	}
	return plan, nil
}

// closeAxis is the arithmetic that makes a factorization TRUE rather than
// plausible: the factors of an axis multiply to its dim, or they overshoot it
// by a declared rounding. Undershooting is the silent one — a layout that
// addresses 3968 of 4000 rows reads 32 rows of another tensor.
func (m *LayoutMorphism) closeAxis(axis int, product uint64, shape Shape) error {
	if axis >= len(shape) {
		return refuse("layout-rank", "%s factors axis %d of a rank-%d tensor",
			m.handle, axis, len(shape))
	}
	if product < shape[axis] {
		return refuse("layout",
			"%s factors axis %d into %d elements, and the axis is %d. A "+
				"factorization that does not reach its axis drops the tail",
			m.handle, axis, product, shape[axis])
	}
	return nil
}

// --- bridging two arrangements ---------------------------------------------

// Bridge is the identity of the map that takes bytes arranged as `from` and
// writes them arranged as `to`.
//
// It is COMPUTED, never authored. Each record is already the map from plain
// row-major; the bridge between any two of them is one record's inverse
// followed by the other's, so a hand-written A->B document would be a second
// place to write down a fact both records already carry. The rendering is what
// a derived tree's manifest names and what its derived digest is taken over,
// so it is an identity and it is stable.
func Bridge(from, to Handle) string {
	return from.String() + ">" + to.String()
}

// ParseBridge reads a bridge id back.
func ParseBridge(text string) (Handle, Handle, error) {
	left, right, found := strings.Cut(text, ">")
	if !found {
		return Handle{}, Handle{}, refuse("bridge", "%q is not <from>><to>", text)
	}
	fromHandle, err := ParseHandle(left)
	if err != nil {
		return Handle{}, Handle{}, err
	}
	toHandle, err := ParseHandle(right)
	if err != nil {
		return Handle{}, Handle{}, err
	}
	return fromHandle, toHandle, nil
}

File: crates/tensorfs-core/tests/layout_morphism.rs
//! AUTO-RATIFICATION, and what it does and does not prove.
//!
//! `ratify` proves a record is a WELL-FORMED map: every source element written
//! exactly once, padding zero, and the bytes that come back through the inverse
//! are the bytes that went in. It cannot prove a well-formed map is the RIGHT
//! one — any bijection round-trips — so the records that describe somebody
//! else's kernel are additionally checked here against an ORACLE written from
//! that kernel's own source, independently of the JSON. Both halves matter and
//! neither substitutes for the other.

use std::collections::BTreeMap;

use tensorfs_core::layout_morphism::{
    LayoutClass, LayoutMorphism, arrangement, catalog, probe_shapes, ratify,
};

fn record(handle: &str) -> &'static LayoutMorphism {
    arrangement(handle).unwrap_or_else(|error| panic!("{handle}: {error}"))
}

/// A shipped record, decoded to JSON so a test can change ONE index and watch
/// the guard fire. The digest is dropped: the Go loader checks it, and here it
/// would refuse every mutation before the guard under test ever ran.
fn mutable(handle: &str) -> serde_json::Value {
    let mut value = serde_json::to_value(SerializableRecord::from(record(handle))).unwrap();
    value.as_object_mut().unwrap().remove("digest");
    value
}

/// The serde shape of a record, for round-tripping one back out to JSON.
#[derive(serde::Serialize)]
struct SerializableRecord {
    format: String,
    name: String,
    version: u32,
    class: &'static str,
    description: String,
    rank: usize,
    sub_axes: Vec<serde_json::Value>,
    permutation: Vec<usize>,
    candidate: bool,
    provenance: Vec<String>,
}

impl From<&LayoutMorphism> for SerializableRecord {
    fn from(source: &LayoutMorphism) -> Self {
        Self {
            format: source.format.clone(),
            name: source.name.clone(),
            version: source.version,
            class: match source.class {
                LayoutClass::Inductor => "inductor",
                LayoutClass::EndpointDeclared => "endpoint-declared",
            },
            description: source.description.clone(),
            rank: source.rank,
            sub_axes: source
                .sub_axes
                .iter()
                .map(|sub| serde_json::json!({"axis": sub.axis, "extent": sub.extent}))
                .collect(),
            permutation: source.permutation.clone(),
            candidate: source.candidate,
            provenance: source.provenance.clone(),
        }
    }
}

fn reparse(value: &serde_json::Value) -> LayoutMorphism {
    serde_json::from_value(value.clone()).expect("a mutated record must still decode")
}

#[test]
fn every_catalogued_arrangement_auto_ratifies() {
    let records = catalog().expect("the vendored catalog must load");
    assert!(
        records.len() >= 7,
        "the v1 vocabulary is closed and has seven entries; the catalog carries {}",
        records.len()
    );
    let mut padded = 0;
    let mut exact = 0;
    for (handle, record) in records {
        let report = ratify(record).unwrap_or_else(|error| panic!("{handle}: {error}"));
        assert!(
            !report.probes.is_empty(),
            "{handle} proved nothing: no probe shape reached it"
        );
        for probe in &report.probes {
            if probe.padded {
                padded += 1;
            } else {
                exact += 1;
            }
        }
        println!("{handle}: {} probe shape(s) ratified", report.probes.len());
    }
    // Both halves of the language have to be exercised, or the padded arm of
    // the verifier is unproven code that happens to compile.
    assert!(exact > 0, "no exact arrangement was ratified");
    assert!(
        padded > 0,
        "no PADDED arrangement was ratified, so round-trip-on-the-image was \
         never the thing under test"
    );
}

#[test]
fn a_permutation_that_is_not_a_bijection_is_refused() {
    let mut document = mutable("torch.channels_last-2d@1");
    // The shipped record ratifies.
    ratify(&reparse(&document)).expect("the shipped record must ratify");
    // ONE index: [0, 2, 3, 1] -> [0, 2, 3, 3].
    document["permutation"] = serde_json::json!([0, 2, 3, 3]);
    let error = ratify(&reparse(&document)).expect_err("a non-bijection was ratified");
    assert!(
        error.to_string().contains("bijection"),
        "the refusal does not name what is wrong: {error}"
    );
}

#[test]
fn a_factorization_that_does_not_reach_its_axis_proves_nothing() {
    let mut document = mutable("cublas.blockscale-128x4@1");
    ratify(&reparse(&document)).expect("the shipped record must ratify");
    // ONE extent: the 32-row factor becomes 16, so the axis addresses half its
    // rows. Every probe shape is then refused — and a record that refuses every
    // probe is a record nothing was proved about, which is a failure and not a
    // pass. This is the arm that catches "the guard never ran".
    document["sub_axes"][2]["extent"] = serde_json::json!("16");
    let error = ratify(&reparse(&document)).expect_err("a short factorization was ratified");
    assert!(
        error.to_string().contains("nothing was proved"),
        "unexpected refusal: {error}"
    );
}

#[test]
fn an_unratified_candidate_never_derives() {
    let mut document = mutable("torch.channels_last-2d@1");
    document["candidate"] = serde_json::json!(true);
    let wish = reparse(&document);
    // It still VERIFIES — a candidate is well-formed data, not garbage.
    ratify(&wish).expect("a candidate must still auto-verify");
    // And it still refuses to derive.
    let error = wish
        .ensure_ratified()
        .expect_err("an unratified candidate was allowed to derive");
    assert!(error.to_string().contains("candidate"), "{error}");
    // Nothing in the shipped catalog is a candidate today; if that changes it
    // is a deliberate act and this line is where it is noticed.
    for (handle, record) in catalog().unwrap() {
        assert!(record.ensure_ratified().is_ok(), "{handle} ships unratified");
    }
}

/// The oracle for `torch.channels_last-2d@1`, written from torch's definition
/// of the memory format rather than from the record: NCHW logical, NHWC in
/// storage. If the record's permutation drifted, these two disagree.
#[test]
fn channels_last_matches_an_independent_nhwc_oracle() {
    let shape = [2u64, 3, 4, 5];
    let plan = record("torch.channels_last-2d@1").plan(&shape).unwrap();
    let (n, c, h, w) = (shape[0], shape[1], shape[2], shape[3]);
    let count = (n * c * h * w) as usize;

    let source: Vec<u8> = (0..count as u32).flat_map(|at| at.to_le_bytes()).collect();
    let mut destination = vec![0u8; count * 4];
    plan.apply(&source, &mut destination, 4).unwrap();

    let mut expected = vec![0u8; count * 4];
    for ni in 0..n {
        for hi in 0..h {
            for wi in 0..w {
                for ci in 0..c {
                    let to = (((ni * h + hi) * w + wi) * c + ci) as usize;
                    let from = (((ni * c + ci) * h + hi) * w + wi) as usize;
                    expected[to * 4..to * 4 + 4]
                        .copy_from_slice(&source[from * 4..from * 4 + 4]);
                }
            }
        }
    }
    assert_eq!(destination, expected, "the record is not torch's channels_last");
}

/// The oracle for `cublas.blockscale-128x4@1`, written from python-gen-worker
/// `nvfp4_quant.to_blocked_scales`' own reshape/permute chain:
///
///   padded.view(nrb, 128, ncb, 4).permute(0, 2, 1, 3)
///        .reshape(-1, 4, 32, 4).transpose(1, 2)
///
/// which puts destination index `(rb*ncb + cb)*512 + b32*16 + a4*4 + c4` at
/// source row `rb*128 + a4*32 + b32`, column `cb*4 + c4`. The record is a
/// transcription of that chain; this test is a second transcription of it,
/// and a transcription slip makes the two disagree.
#[test]
fn the_blocked_scale_record_matches_the_packer_it_was_transcribed_from() {
    // Deliberately RAGGED: 100 rows and 3 block-columns both round up, so the
    // padding path is what is compared, not just the aligned interior.
    let (rows, cols) = (100u64, 3u64);
    let plan = record("cublas.blockscale-128x4@1")
        .plan(&[rows, cols])
        .unwrap();
    let count = (rows * cols) as usize;
    let source: Vec<u8> = (0..count as u32)
        .flat_map(|at| (at + 1).to_le_bytes())
        .collect();
    let mut destination = vec![0u8; plan.dest_elements() as usize * 4];
    plan.apply(&source, &mut destination, 4).unwrap();

    let nrb = rows.div_ceil(128);
    let ncb = cols.div_ceil(4);
    let mut expected = vec![0u8; (nrb * 128 * ncb * 4) as usize * 4];
    for rb in 0..nrb {
        for cb in 0..ncb {
            for a4 in 0..4u64 {
                for b32 in 0..32u64 {
                    for c4 in 0..4u64 {
                        let to = ((rb * ncb + cb) * 512 + b32 * 16 + a4 * 4 + c4) as usize;
                        let row = rb * 128 + a4 * 32 + b32;
                        let col = cb * 4 + c4;
                        if row < rows && col < cols {
                            let from = (row * cols + col) as usize;
                            expected[to * 4..to * 4 + 4]
                                .copy_from_slice(&source[from * 4..from * 4 + 4]);
                        }
                    }
                }
            }
        }
    }
    assert_eq!(
        destination.len(),
        expected.len(),
        "the record and the packer disagree on the destination size"
    );
    assert_eq!(
        destination, expected,
        "the record does not reproduce to_blocked_scales"
    );
}

/// THE RED ARM OF BOTH ORACLES: a record that is still a perfect bijection —
/// so `ratify` accepts it, and must — but is the WRONG permutation. This is
/// exactly the failure auto-ratification cannot see and the reason these
/// records carry a cited source and an oracle test.
#[test]
fn a_bijective_but_wrong_permutation_survives_ratification_and_fails_the_oracle() {
    let mut document = mutable("cublas.blockscale-128x4@1");
    // Swap the two 4-extent sub-axes' storage positions: [0,3,2,1,4] ->
    // [0,3,2,4,1]. Still a bijection; still round-trips; different bytes.
    document["permutation"] = serde_json::json!([0, 3, 2, 4, 1]);
    let wrong = reparse(&document);
    ratify(&wrong).expect(
        "a wrong-but-bijective permutation MUST still auto-ratify — if this \
         fails, the claim that ratification proves correctness is being made \
         somewhere it should not be",
    );

    let (rows, cols) = (100u64, 3u64);
    let plan = wrong.plan(&[rows, cols]).unwrap();
    let right = record("cublas.blockscale-128x4@1").plan(&[rows, cols]).unwrap();
    let count = (rows * cols) as usize;
    let source: Vec<u8> = (0..count as u32)
        .flat_map(|at| (at + 1).to_le_bytes())
        .collect();
    let mut mangled = vec![0u8; plan.dest_elements() as usize * 4];
    let mut correct = vec![0u8; right.dest_elements() as usize * 4];
    plan.apply(&source, &mut mangled, 4).unwrap();
    right.apply(&source, &mut correct, 4).unwrap();
    assert_ne!(
        mangled, correct,
        "two different permutations produced the same bytes, so the oracle \
         tests above are not measuring the permutation"
    );
}

/// The cross-LANGUAGE guard. Go decides with these records and this crate moves
/// the bytes with them, which is two evaluators of one shape language. They are
/// held together by banked vectors: this test writes them, the Go suite checks
/// them, and a drift in either evaluator fails on one side or the other.
#[test]
fn plan_vectors_agree_with_the_bank() {
    let path = concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../../spec/v2/vectors/layout-plans.json"
    );
    let mut plans = Vec::new();
    for (handle, record) in catalog().unwrap() {
        for shape in probe_shapes(record) {
            let Ok(plan) = record.plan(&shape) else {
                continue;
            };
            plans.push(serde_json::json!({
                "layout": handle,
                "shape": shape,
                "source_elements": plan.source_elements(),
                "dest_elements": plan.dest_elements(),
                "padded": plan.padded(),
                "storage_extents": plan.storage_extents(),
            }));
        }
    }
    let banked = serde_json::json!({
        "format": "tensorfs-layout-plan-vectors-v1",
        "note": "Generated by crates/tensorfs-core/tests/layout_morphism.rs. \
                 Checked by the Go suite against its own evaluator: these are the \
                 two implementations of one shape language agreeing in writing.",
        "plans": plans,
    });
    let rendered = serde_json::to_string_pretty(&banked).unwrap() + "\n";
    if std::env::var("TENSORFS_REBANK").is_ok() {
        std::fs::write(path, &rendered).expect("rebank layout-plans.json");
        return;
    }
    let existing = std::fs::read_to_string(path).unwrap_or_else(|error| {
        panic!("{path}: {error}. Run with TENSORFS_REBANK=1 to write it.")
    });
    let left: serde_json::Value = serde_json::from_str(&existing).unwrap();
    assert_eq!(
        left, banked,
        "the banked plan vectors and this crate's evaluator disagree"
    );
}

/// Every record's probe set is derived from the record, so this is a guard on
/// the DERIVATION: a vocabulary entry whose alignment is misread would silently
/// be probed at a shape that proves less.
#[test]
fn probe_shapes_are_derived_and_cover_both_arms() {
    let mut shapes = BTreeMap::new();
    for (handle, record) in catalog().unwrap() {
        let probes = probe_shapes(record);
        assert!(!probes.is_empty(), "{handle} has no probe shape");
        for probe in &probes {
            if record.rank != 0 {
                assert_eq!(
                    probe.len(),
                    record.rank,
                    "{handle}: a rank-{} record was probed at {probe:?}",
                    record.rank
                );
            }
        }
        shapes.insert(handle.clone(), probes);
    }
    // The exact-division records must refuse their ragged probe; the padded
    // ones must accept it. If every record accepted every probe, the "skip a
    // refused shape" path in `ratify` would be dead code.
    let exact = record("nunchaku.micro-scale@1");
    let ragged = shapes["nunchaku.micro-scale@1"].last().unwrap();
    assert!(
        exact.plan(ragged).is_err(),
        "an exact factorization accepted a ragged shape {ragged:?}"
    );
    let padded = record("cublas.blockscale-128x4@1");
    let ragged = shapes["cublas.blockscale-128x4@1"].last().unwrap();
    assert!(
        padded.plan(ragged).is_ok(),
        "a padded factorization refused a ragged shape {ragged:?}"
    );
}
